### PR TITLE
enh: Text Filter waits until user finishes updating

### LINKED ISSUE #2112: Text filter should wait until user finishes before updating.
**Is your feature request related to a problem or area of OpenRefine? Please describe.**
When entering a string or regex expression in the text filter, I'm often prevented from finishing the expression before OpenRefine attempts to update the results. This is frustrating, especially when trying to formulate complex regex.

**Describe the solution you'd like**
My preferred solution would be a `Submit` button (ideally with a keyboard shortcut) so that OpenRefine doesn't keep trying to update while I'm typing.

**Describe alternatives you've considered**
The only alternative I've found is to write regex expressions in a text file and copy/paste into the text filter.

### PR DESCRIPTION
**Pull request by kushthedude**
Signed-off-by: kushthedude <kushthedude@gmail.com>

Fixes #2112

### CODE REVIEW

**Comment by kushthedude** [comment]
@wetneb I was thinking of having PR Template for the Open Refine. Please tell me if it is good to go, I will go ahead and make PR

File: main/webapp/modules/core/scripts/facets/text-search-facet.js [hunk 1]
```diff
@@ -147,7 +147,8 @@ TextSearchFacet.prototype._initializeUI = function() {
       return;
     }
     self._query = this.value;
-    self._scheduleUpdate();
+    setTimeout(function(){ self._scheduleUpdate(); }, 1500);
```
Comment (thadguidry) [inline main/webapp/modules/core/scripts/facets/text-search-facet.js]: 1200 and 1300 feel a bit more natural to me, whereas 1500+ feels awkward. But then, the original issue was that of Regex entry into the Text Filter box, and when I test with a bit of Regex, indeed, 1500 is a good balance. Let's go with that!

Comment (kushthedude) [reply main/webapp/modules/core/scripts/facets/text-search-facet.js]: So it's good to go then ?

Comment (thadguidry) [reply main/webapp/modules/core/scripts/facets/text-search-facet.js]: We'll let a few other folks who use this often and see what they think.  It seems to work fine also and delay with the Regex error propagation also for instance if doing a `(a)(b)`

@ostephens @ettorerizza How does this feel for you when typing Regex?  too long? too short?

**Comment by kushthedude** [comment]
Cool I will make the changes in the _scheduleUpdate function.

**Comment by kushthedude** [comment]
@wetneb Done, as there was timeout of 500ms already and I added timeout of 1500ms additionally which was working quite fine. Hence I increased the original timeout to 2000. Please let me know if additional changes are required.

**Comment by kushthedude** [comment]
@ostephens @thadguidry I have made separated timeout for both types of filters. I used the Reg-Ex and Case-Sensitive filters both, it works quite great.

**Comment by wetneb** [comment]
Yes absolutely. At the moment the proposed change has two issues, in terms of actual behaviour:
* adding delays when boxes are checked;
* keeping the same 500ms delay on keystrokes, regardless of whether the regex checkbox is enabled.
The right place where to add this differentiated delay is indeed in the `_scheduleUpdate` function.

**Comment by ostephens** [comment]
@kushthedude are you happy to make changes in line with the comment from @wetneb and me?

**Comment by kushthedude** [comment]
I was working on recon issue meanwhile that's why is the delay. I will
complete it by tujhe.

**Comment by kushthedude** [comment]
@wetneb @ostephens @thadguidry I tried to implement a timeout in the schedule_update function itself. Please have a look timeout is same for every keystroke and for case-sensitive/regex differential values have been set for an additional timeout.

**Comment by kushthedude** [comment]
> At the moment there is a 500 ms timeout followed by either a 250 ms timeout or a 1000 ms timeout. Wouldn't it be cleaner to just have one timeout, whose length would vary depending on the checkbox's state? So, vary the existing 500 ms timeout instead of introducing another layer of timeouts, I would say.

@wetneb I could have used the ternary operator if we had two different types of time-outs but we have 3 timeouts right now:
- 500 ms for key-stroke.
- 750 ms for Case-Sensitive
- 1500 ms for RegEx.
Can you suggest a better way to utilise the same?

**Comment by thadguidry** [comment]
@kushthedude Make a single timeout object which takes a delay of ms.   Based on if the user checks some boxes... set the appropriate delay in ms.  We don't want 3 timeout objects.  In other words, the input box has 1 single timeout... what the timeout delay is...will be based on criteria the user clicked on various boxes.  Does that make more sense?

**Comment by kushthedude** [comment]
> @kushthedude Make a single timeout object which takes a delay of ms. Based on if the user checks some boxes... set the appropriate delay in ms. We don't want 3 timeout objects. In other words, the input box has 1 single timeout... what the timeout delay is...will be based on criteria the user clicked on various boxes. Does that make more sense?

Damn, I didn't get the idea to create an object. Thanks, @thadguidry for it!

**Comment by thadguidry** [comment]
Cool.  And if you read above, it doesn't have to be an object.  As @ostephens states, it could just be you change the way the higher level function works.  `TextSearchFacet.prototype._scheduleUpdate` function.

And I agree with @ostephens the only criteria for now would be the Regex checkbox that increases or decreases the delay.  The case sensitive box shouldn't affect anything.

OK, now get to work and figure it all out! :-)

**Comment by kushthedude** [comment]
> And I agree with @ostephens the only criteria for now would be the Regex checkbox that increases or decreases the delay. The case sensitive box shouldn't affect anything.

@thadguidry Please have a look now. Since the only Regex should vary the timeout, I didn't need to make a seperate variable for a timeout.

**Comment by kushthedude** [comment]
> That's much better (and simpler!). If we agree on the specific values of the timeouts, then it works for me.

I think 1500ms for regex was right for everyone as can be seen from the above discussion. @ostephens @thadguidry What are your views on the same?

**Comment by thadguidry** [comment]
1000 - 1500 is fine.  My preference seemed to be around 1200.  But up to you.

**Comment by kushthedude** [comment]
> 1000 - 1500 is fine. My preference seemed to be around 1200. But up to you.

I think 1500 is fine as regex are quite complex than typing letter. But still, if 1200 is what all want I have no object for it.

**Comment by thadguidry** [comment]
This will never end.
I think what the world wants here, especially with Regex, is that it should be a button to submit (or Enter key), and not auto-submitting in all honesty, since it could be 1 million rows to process or more eventually.  Some users will want 1.5 secs, some will want 5 or 10 secs.  A submit or Go button that is mapped to the ENTER key upon focus inside the input box would be the best of all, just like the original poster (OP) asked for in #2112.  The flip side is having a Cancel button for the processing (but this would be harder to implement and not as user friendly.)  A button keybinded will give the power back to the user, instead of us making a decision for him on how many seconds.  [Our core philosophy.](https://github.com/OpenRefine/OpenRefine/pull/2366/files)

**Comment by kushthedude** [comment]
Got it @thadguidry . I will be making changes as per you asked !

**Comment by thadguidry** [comment]
@kushthedude I'd show it as a the FILTER button (with ENTER keybinding) since it is a Text Filter operation ... something like this pretty picture :-)  I think this button should show always with the Text Filter box and would lend clarity to what the darn thing actually does!

image

**Comment by wetneb** [comment]
@thadguidry hang on, that's a completely different change from what @kushthedude has been working on so far… It's not great to have him do many changes in the way timeouts are defined, just to change our mind at the last minute to switch to a totally different solution, no?

I would be inclined to merge this as such - replacing the timeout by a button is something we can discuss in another issue, as it brings other complications (how do you display a text filter that has been changed and not validated, for instance?)

**Comment by thadguidry** [comment]
@wetneb absolutely.

**Comment by kushthedude** [comment]
@thadguidry @wetneb What I suggest for the now is let's move ahead with this PR and see how it works for the users. But if the users still need a seperate text filter button after this, we can cover it in a cover-up PR.
Tell me how it sounds to you? :)

**Comment by wetneb** [comment]
@thadguidry thanks, yes I am not opposed to a button at all.

**Comment by thadguidry** [comment]
@kushthedude yes

**Comment by ostephens** [comment]
Thanks @kushthedude - great!

**Pull request merged**